torch/_functorch/_aot_autograd/utils.py

```diff
@@ -124,17 +124,18 @@ class PytreeThunk:
     )
     is_really_simple = None  # if the output spec is a LeafSpec
 
-    def set(self, spec):
+    def set(self, spec: pytree.TreeSpec) -> None:
         assert self.spec is None or self.spec == spec
-        self.spec = spec
-        if type(self.spec) in [tuple, list] and all(
-            isinstance(i, pytree.LeafSpec) for i in spec.children_specs
+        assert spec is not None
+        self.spec: pytree.TreeSpec = spec
+        if type(self.spec) in {tuple, list} and all(
+            child.is_leaf() for child in spec.children_specs
         ):
             self.is_simple = True
-        if isinstance(self.spec, pytree.LeafSpec):
+        if self.spec.is_leaf():
             self.is_really_simple = True
 
-    def unflatten(self, x):
+    def unflatten(self, x: List[Any]) -> Any:
```

Comment:
You could improve with a TypeVar, no?
```suggestion
    def unflatten(self, x: List[_T]) -> _T:
```

Reply:
Actually, wow all the flatten / unflatten functions in pytree are really poorly typed. They should try to retain the typing in the iterable if possible like itertools does.

Reply:
The input type is `List[T]`, while the output type can be `T | List[T] | PyTree[T]`. Currently the annotation for `PyTree` is `Any`.